plugins/protocols/session-lock.cpp

```diff
@@ -0,0 +1,434 @@
+#include "wayfire/core.hpp"
+
+#include "wayfire/seat.hpp"
+#include "wayfire/scene-input.hpp"
+#include "wayfire/scene-operations.hpp"
+#include "wayfire/unstable/wlr-surface-node.hpp"
+#include "wayfire/unstable/wlr-surface-controller.hpp"
+#include "wayfire/output.hpp"
+#include "wayfire/output-layout.hpp"
+#include "wayfire/signal-definitions.hpp"
+
+#include <wayfire/plugin.hpp>
+#include <wayfire/util.hpp>
+#include <wayfire/nonstd/wlroots-full.hpp>
+#include <wayfire/plugins/common/simple-text-node.hpp>
+#include <wayfire/util/log.hpp>
+
+enum lock_state
+{
+    LOCKING,
+    LOCKED,
+    UNLOCKED,
+    DESTROYED,
+    ZOMBIE,
+} state;
+
+class wf_session_lock_plugin : public wf::plugin_interface_t
+{
+    class wayfire_session_lock;
+
+    class lock_surface_node_t : public wf::scene::wlr_surface_node_t
+    {
+      public:
+        lock_surface_node_t(wayfire_session_lock *lock,
+            wlr_session_lock_surface_v1 *lock_surface,
+            wf::output_t *output) :
+            wf::scene::wlr_surface_node_t(lock_surface->surface, true /* autocommit */),
+            lock(lock),
+            lock_surface(lock_surface),
+            output(output),
+            interaction(std::make_unique<lock_surface_keyboard_interaction>(lock_surface->surface,
+                lock_surface->output))
+        {
+            lock_surface_destroy.set_callback([this] (void *data)
+            {
+                wf::wlr_surface_controller_t::try_free_controller(this->lock_surface->surface);
+                wf::scene::remove_child(shared_from_this());
+                lock_surface_destroy.disconnect();
+                this->lock->surface_destroyed(this->output);
+                const char *name = this->output->handle ? this->output->handle->name : "(deleted)";
+                LOGD("lock_surface on ", name, " destroyed");
```

Comment:
You can/should reset the keyboard interaction to something else (something like this: https://github.com/WayfireWM/wayfire/blob/master/src/view/wlr-surface-node.cpp#L102) so that we don't get a crash if someone holds on to the node after the surface/output are destroyed.

Reply:
And also re lifetimes, you should set lock_surface to NULL here, and make sure it is not used in the other functions if it is NULL already (where it makes sense)

Reply:
Attempted to reset the keyboard interaction.

I don't think lock_surface can be used after it's destroyed. The only other method that uses it is attach_to_layer. But attach_to_layer will only be called if the output_state->surface is non-null. But when the lock_surface is destroyed, the callback for the destroy event calls this->lock->surface_destroyed, which nulls out the output_state->surface, so attach_to_layer will not be called.

Reply:
Ok, it might still be useful to set it to NULL so that if in the future we change anything, we directly get a null pointer crash instead of a dangling pointer.